protocols/autonatv2/src/client/behaviour.rs

```diff
@@ -0,0 +1,424 @@
+                        self.pending_events
+                            .push_back(ToSwarm::ExternalAddrExpired(addr.clone()));
+                    }
+                    dial_request::InternalError::InternalServer
+                    | dial_request::InternalError::DataRequestTooLarge { .. }
+                    | dial_request::InternalError::DataRequestTooSmall { .. }
+                    | dial_request::InternalError::InvalidResponse
+                    | dial_request::InternalError::ServerRejectedDialRequest
+                    | dial_request::InternalError::InvalidReferencedAddress { .. }
+                    | dial_request::InternalError::ServerChoseNotToDialAnyAddress => {
+                        self.handle_no_connection(peer_id, connection_id);
+                    }
+                    _ => {
+                        tracing::debug!("Test failed: {:?}", err);
+                    }
+                }
+            }
+            Either::Left(dial_request::ToBehaviour::StatusUpdate(update)) => self
+                .pending_events
+                .push_back(ToSwarm::GenerateEvent(update)),
+        }
+    }
+
+    fn poll(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<ToSwarm<Self::ToSwarm, <Self::ConnectionHandler as ConnectionHandler>::FromBehaviour>>
+    {
+        if let Some(event) = self.pending_events.pop_front() {
+            return Poll::Ready(event);
+        }
+        if self.next_tick.poll_unpin(cx).is_ready() {
+            self.inject_address_candiate_test();
+            if let Some(event) = self.pending_events.pop_front() {
+                return Poll::Ready(event);
+            }
+        }
+        Poll::Pending
+    }
+}
+
+impl<R> Behaviour<R>
+where
+    R: RngCore + 'static,
+{
+    pub fn new(rng: R, config: Config) -> Self {
+        Self {
+            local_peers: HashSet::new(),
+            pending_nonces: HashSet::new(),
+            known_servers: Vec::new(),
+            rng,
+            next_tick: Delay::new(config.recheck_interval),
+            config,
+            pending_events: VecDeque::new(),
+            address_candidates: HashMap::new(),
+            peers_to_handlers: HashMap::new(),
+            already_tested: HashSet::new(),
+        }
+    }
+
+    /// Injects a known server into the behaviour. It's mostly useful if you are not using identify
+    /// or for testing purposes.
+    pub fn inject_kown_server(&mut self, peer: PeerId) {
+        if !self.known_servers.contains(&peer) {
+            self.known_servers.push(peer);
+        }
+    }
+
+    /// Inject a new address candidate into the behaviour.
+    pub fn inject_address_candiate(&mut self, addr: Multiaddr) {
+        *self.address_candidates.entry(addr).or_default() += 1;
+    }
+
+    /// Inject an immediate test for all pending address candidates.
+    pub fn inject_address_candiate_test(&mut self) {
+        if self.known_servers.is_empty() || self.address_candidates.is_empty() {
```

Comment:
The 2nd check seems obsolete to me because it is covered by the check on line 287.

Reply:
It's now different, once I pushed the changes, please check if it's still redundant.